man/jsonlogalert.1
.TH "JSONLOGALERT" "1" "2024-10-02" "1.0.0.dev0" "jsonlogalert Manual"
.SH NAME
jsonlogalert \- Read JSON structured logs and filter...
.SH SYNOPSIS
.B jsonlogalert
[OPTIONS] [LOGFILE]...
.SH DESCRIPTION
Read JSON structured logs and filter entries for unusual activity.
.PP
    When no LOGFILE(s) are provided, read/tail and filter logs for configured log sources.
    Alternatively, read and/or tail and filter LOGFILE(s) for a specified SOURCE.
    When LOGFILE is -, read standard input.
    
.SH OPTIONS
.TP
\fB\-c,\fP \-\-config\-file FILE
Read options from configuration FILE.  [default: /etc/jsonlogalert.conf]
.TP
\fB\-d,\fP \-\-config\-dir DIRECTORY
Set path to directory containing source and service definitions.  [default: /etc/jsonlogalert.d]
.TP
\fB\-v,\fP \-\-verbose
Be more verbose; can specify more than once.
.TP
\fB\-\-version\fP
Show the version and exit.
.TP
\fB\-s,\fP \-\-source SOURCE
Read only this log SOURCE; can specify more than once.
.TP
\fB\-\-service\fP SERVICE
Enable only this SERVICE for a log SOURCE; can specify more than once.
.TP
\fBTAIL\fP LOG OPTIONS:
.PP
.TP
\fB\-J,\fP \-\-journal\-dir DIRECTORY
Tail systemd journal DIRECTORY.
.TP
\fB\-\-tail\-journal\-bin\fP PATH
Set path of executable to tail systemd journal.  [default: logtail-journal]
.TP
\fB\-\-tail\-journal\-since\fP [today|boot|all]
Read all systemd journal entries or since last boot (ignores cursor.)
.TP
\fB\-f,\fP \-\-tail\-file LOGFILE
Tail text log LOGFILE; can specify more than once.
.TP
\fB\-\-tail\-file\-bin\fP PATH
Set path of executable to tail log files.  [default: logtail2]
.TP
\fB\-\-tail\-state\-dir\fP DIRECTORY
Set path of DIRECTORY to save tail offset/cursor state.  [default: /var/lib/misc]
.TP
\fB\-\-tail\-reset\fP
Delete offset/cursor state files and exit.
.TP
\fB\-\-tail\-dryrun\fP
Use but do not update tail offset/cursor.
.TP
\fB\-\-tail\-ignore\fP
Ignore and do not update tail offset/cursor.
.TP
\fBGENERAL\fP OUTPUT OPTIONS:
.PP
.TP
\fB\-\-output\-stdout\fP
Output results to stdout.
.TP
\fB\-\-output\-devnull\fP
Output results to /dev/null.
.TP
\fB\-\-output\-template\-file\fP FILENAME
Use FILENAME instead of default output template.
.TP
\fB\-\-max\-logentries\fP INTEGER
Maximum number of entries to report.  [default: 250]
.TP
\fBFILE\fP OUTPUT OPTIONS:
.PP
.TP
\fB\-\-output\-file\-dir\fP DIRECTORY
Output results to file in DIRECTORY.
.TP
\fB\-\-output\-file\-name\fP FILENAME
Output results to FILENAME in DIRECTORY when a single SERVICE is specified.
.TP
\fBSMTP\fP OUTPUT OPTIONS:
.PP
.TP
\fB\-\-output\-smtp\fP / \-\-no\-output\-smtp
Enable SMTP output.  [default: output-smtp]
.TP
\fB\-\-output\-smtp\-host\fP HOSTNAME
Mail server hostname or address.  [default: localhost]
.TP
\fB\-\-output\-smtp\-port\fP INTEGER
Mail server port.  [default: 25]
.TP
\fB\-\-output\-smtp\-auth\-ssl\fP
Mail server uses SSL connection.
.TP
\fB\-\-output\-smtp\-auth\-tls\fP
Mail server uses TLS.
.TP
\fB\-\-output\-smtp\-auth\-username\fP USERNAME
Mail server authentication username.
.TP
\fB\-\-output\-smtp\-auth\-password\fP PASSWORD
Mail server authentication password.
.TP
\fB\-\-output\-smtp\-rcpt\fP EMAIL
Email recipient address. Required.
.TP
\fB\-\-output\-smtp\-sender\fP EMAIL
Email sender address.  [default: recipient address]
.TP
\fB\-\-output\-smtp\-rcpt\-name\fP NAME
Email recipient name.
.TP
\fB\-\-output\-smtp\-sender\-name\fP NAME
Email sender name.  [default: recipient name]
.TP
\fB\-\-output\-smtp\-subject\fP SUBJECT
Email subject line.
.TP
\fBTIMING\fP OPTIONS:
.PP
.TP
\fB\-\-wait\fP SECONDS
Wait SECONDS before scanning (give logs time to settle.)
.TP
\fB\-\-slowroll\fP SECONDS
Watch for log activity for no less than SECONDS (to satisfy rate limits.)
.TP
\fB\-\-batch\-interval\fP SECONDS
Watch for more log activity within SECONDS.
.TP
\fBDIAGNOSTIC\fP OPTIONS:
.PP
.TP
\fB\-\-dry\-run\fP
Run without using or updating tail offset/cursor.
.TP
\fB\-\-print\-rules\fP
Print rules and exit.
.TP
\fB\-\-print\-conf\fP
Print source and service configurations and exit.
.TP
\fB\-\-print\-field\-types\fP
Print source field types and exit.
.TP
\fB\-\-skip\-drop\-rules\fP
Skip drop rules.
.TP
\fB\-\-skip\-capture\-fields\fP
Capture all fields.
.TP
\fB\-\-skip\-conceal\-fields\fP
Do not conceal fields.
.TP
\fB\-\-skip\-ignore\-fields\fP
Do not ignore fields.


jsonlogalert.sh
#!/usr/bin/env sh

# systemd services *will not* run scripts that do not have the appropriate SELinux context.
# This undoubtably means we cannot run 'jsonlogalert' directly from the virtual environment.
# This script is a stub that can be placed in a directory where it can inherit the required context.

set -u

# 'install.sh' hard-codes VENVDIR to the virtual environment path;
# Defaulting to VIRTUAL_ENV is a dev environment fallback.
VIRTUAL_ENV="${VIRTUAL_ENV:-}"
VENVDIR=
VENVDIR="${VENVDIR:-$VIRTUAL_ENV}"
SCRIPT="$VENVDIR"/bin/jsonlogalert
SYSTEMD_ETC=/etc/systemd/system

# Print usage and exit
usage()
{
	cat <<EOF
Try 'jsonlogalert --help' for more help.

USAGE
  jsonlogalert systemd COMMAND
  jsonlogalert [OPTIONS] [LOGFILE]...

SYNOPSIS
  systemd      Convenient way to manage systemd service, timer and path units
               with the prefix 'jsonlogalert-'.

COMMANDS
  list         List timer and path units.
  enable       Enable and start timer and path units.
  disable      Disable and stop timer and path units.
  install DIR  Copy units from DIR to '$SYSTEMD_ETC'.
  uninstall    Remove units from '$SYSTEMD_ETC'.

OPTIONS
  -h    Show this help and exit
EOF
	exit 2
}

# Write to stderr; use `echo` for stdout.
msg()
{
	echo >&2 "$@"
}

# Write to stderr and exit non-zero
die()
{
	message="${1:-}"
	[ -z "$message" ] || msg "$message"
	exit 1
}

###############################################################################
# main

# Parse command line options
while getopts ":h-:" opt; do
	case "$opt" in
		h) usage ;;
        *);;
	esac
done

# Convenience option to manage systemd timer and path units.
if [ -n "${1:-}" ] && [ "$1" = "systemd" ]; then
    systemd_cmd="${2:-}"
    systemd_opt="${3:-}"

    if [ "$systemd_cmd" = "list" ]; then
        find "$SYSTEMD_ETC" -maxdepth 1 -regextype egrep -regex '.*/jsonlogalert-[^.]+\.(path|timer)' -printf "%f\n"
    elif [ "$systemd_cmd" = "enable" ] || [ "$systemd_cmd" = "disable" ]; then
        find "$SYSTEMD_ETC" -maxdepth 1 -regextype egrep -regex '.*/jsonlogalert-[^.]+\.(path|timer)' -printf "%f\n" | xargs --no-run-if-empty --verbose systemctl --quiet --no-reload "$systemd_cmd" --now
        systemctl daemon-reload
    elif [ "$systemd_cmd" = "uninstall" ]; then
        find "$SYSTEMD_ETC" -maxdepth 1 -regextype egrep -regex '.*/jsonlogalert-[^.]+\.(path|timer|service)' -delete
    elif [ "$systemd_cmd" = "install" ] && [ -n "$systemd_opt" ] && [ -d "$systemd_opt" ]; then
        find "$systemd_opt" -maxdepth 1 -regextype egrep -regex '.*/jsonlogalert-[^.]+\.(path|timer|service)' -print0 | xargs -0 --no-run-if-empty --verbose cp -t "$SYSTEMD_ETC"
    else
        usage
    fi

    exit 0
fi

[ -n "$VENVDIR" ] || die "This script is a stub; it is not standalone until install.sh copies it; you can also set VIRTUAL_ENV"
[ -d "$VENVDIR" ] || die "$VENVDIR: No such directory."
[ -f "$SCRIPT"  ] || die "$SCRIPT: No such file."

exec "$SCRIPT" "$@"
